### PR TITLE
Feature/bi 8434 valid char report validation

### PR DESCRIPTION
**Pull request by CH-JonathanBlake**
https://companieshouse.atlassian.net/browse/BI-8434

- Applied Charset2 validation to several existing validation methods.
- Created additional validation methods
- Applied optional validation methods to the Creation Validation method
- Restructured tests.

### CODE REVIEW

File: src/test/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidatorUnitTest.java [hunk 1]
```diff
@@ -63,462 +71,632 @@ void setUp() {
...
+                Errors errors = new Errors();
+                PscDiscrepancyReport pscDiscrepancyReport = new PscDiscrepancyReport();
+                pscDiscrepancyReport.setObligedEntityType(VALID_TYPE);
+
+                Errors errorsFromValidation =
+                        pscDiscrepancyReportValidator.validateForCreation(pscDiscrepancyReport, errors);
+
+                ObjectMapper mapper = new ObjectMapper();
+                try {
+                    System.out.println(mapper.writeValueAsString(errorsFromValidation));
```
Comment (Zain-Abbas1) [inline src/test/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidatorUnitTest.java:93]: Can we remove this system.out print line please

File: src/main/java/uk/gov/ch/pscdiscrepanciesapi/validation/Validators.java [hunk 2]
```diff
@@ -15,11 +15,14 @@
...
 
-    public boolean validateNotBlank(String toTest, String location, Errors errs) {
-        if (validateNotNull(toTest, location, errs)) {
-            return validateNotEmpty(toTest, location, errs);
+    public void nullOrValidate(Object nullable, Runnable validationCheck) {
+        if(nullable != null){
+            validationCheck.run();
         }
-        return false;
+    }
```
Comment (Zain-Abbas1) [inline src/main/java/uk/gov/ch/pscdiscrepanciesapi/validation/Validators.java:22]: Can we update the unit tests for this change please

Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/pscdiscrepanciesapi/validation/Validators.java:22]: I don't know why Github has presented this as a change in this way, validateNotBlank is below and logically identical. 
Will add tests for nullOrValidate.

File: src/test/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidatorUnitTest.java (hunk 1 above)
Comment (CH-JonathanBlake) [reply src/test/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidatorUnitTest.java]: Done.

File: src/main/java/uk/gov/ch/pscdiscrepanciesapi/models/rest/ReportStatus.java [hunk 3]
```diff
@@ -7,4 +7,13 @@
...
     FAILED_TO_SUBMIT;
+
+    public static boolean isValid(String status) {
+        for(ReportStatus rs : ReportStatus.values()){
+            if(rs.name().equals(status)){
+                return true;
+            }
+        }
+        return false;
+    }
```
Comment (Zain-Abbas1) [inline src/main/java/uk/gov/ch/pscdiscrepanciesapi/models/rest/ReportStatus.java:18]: Do we have test class for this?

File: src/main/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidator.java [hunk 4]
```diff
@@ -18,41 +14,35 @@
...
     private static final String OBLIGED_ENTITY_EMAIL = "obliged_entity_email";
     private static final String OBLIGED_ENTITY_CONTACT_NAME = "obliged_entity_contact_name";
     private static final String OBLIGED_ENTITY_ORGANISATION_NAME = "obliged_entity_organisation_name";
     private static final String OBLIGED_ENTITY_TYPE = "obliged_entity_type";
     private static final String COMPANY_INCORPORATION_NUMBER = "company_number";
     private static final String STATUS = "status";
     private static final String ETAG = "etag";
+    private static final String OBLIGED_ENTITY_NAME = "obliged_entity_name";
+    private static final String SUBMISSION_REFERENCE = "submission_reference";
+    public static final String INVALID_CHAR_MESSAGE = " contains an invalid character";
```
Comment (Zain-Abbas1) [inline src/main/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidator.java:27]: Any reason why this is public?

File: src/main/java/uk/gov/ch/pscdiscrepanciesapi/models/rest/ReportStatus.java (hunk 3 above)
Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/pscdiscrepanciesapi/models/rest/ReportStatus.java:18]: I will add them.

File: src/main/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidator.java (hunk 4 above)
Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/pscdiscrepanciesapi/validation/PscDiscrepancyReportValidator.java]: changed.

File: src/main/java/uk/gov/ch/pscdiscrepanciesapi/models/rest/ReportStatus.java (hunk 3 above)
Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/pscdiscrepanciesapi/models/rest/ReportStatus.java:18]: I have added around 10,010 tests to attempt to cover most scenarios.

**Review by Zain-Abbas1 — APPROVED** [review]

**Review by ch-rfishwick — APPROVED** [review]

**Pull request merged**